forgotten-sender/listing-04-fix-2/main.go

```diff
@@ -0,0 +1,70 @@
+// Example program with the goroutine leak fixed. The code cancels the worker
+// on its way out of the process function.
+package main
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+)
+
+func main() {
+
+	// Report number of goroutines. Will be 1.
+	fmt.Println("Number of goroutines:", runtime.NumGoroutine())
+
+	process(context.Background())
+
+	// Hold the program from terminating for 1 second to see if the goroutine
+	// created by process will terminate.
+	time.Sleep(time.Second)
+
+	// Report number of goroutines. Will be 1.
+	fmt.Println("Number of goroutines:", runtime.NumGoroutine())
+}
+
+func process(ctx context.Context) {
+
+	// Wrap the parent context in a new context with a 100ms deadline. The
+	// context will be canceled if 100ms passes or if the parent function's
+	// context is canceled for any reason.
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
```

Comment:
This timeout should come from the caller. We need to teach that higher levels calls tell lower level calls how long they want to wait.